swupd-inspector/diff.go

```diff
@@ -0,0 +1,268 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"path/filepath"
+	"sort"
+
+	"github.com/clearlinux/mixer-tools/internal/client"
+	"github.com/clearlinux/mixer-tools/swupd"
+)
+
+// TODO: MoM signature verification.
+
+// Terminal colors.
+var (
+	RED   = "\x1b[31m"
+	GREEN = "\x1b[32m"
+	RESET = "\x1b[0m"
+)
+
+func runDiff(cacheDir string, args []string) {
+	var (
+		noColor bool
+		strict  bool
+	)
+
+	flagSet := flag.NewFlagSet("diff", flag.ExitOnError)
+	flagSet.BoolVar(&noColor, "no-color", false, "use no colors")
+	flagSet.BoolVar(&strict, "strict", false, "also compare version numbers")
+	flagSet.Usage = usage
+	err := flagSet.Parse(args)
+	if err != nil {
+		log.Fatalf("ERROR: %s", err)
+	}
+
+	if len(flagSet.Args()) != 2 {
+		usage()
+		return
+	}
+
+	if noColor {
+		RED = ""
+		GREEN = ""
+		RESET = ""
+	}
+
+	baseA, versionA := parseURL(flagSet.Arg(0))
+	baseB, versionB := parseURL(flagSet.Arg(1))
+
+	var stateA, stateB *client.State
+	stateDirA := filepath.Join(cacheDir, convertContentBaseToDirname(baseA))
+	stateA, err = client.NewState(stateDirA, baseA)
+	if err != nil {
+		log.Fatalf("ERROR: %s", err)
+	}
+
+	stateDirB := filepath.Join(cacheDir, convertContentBaseToDirname(baseB))
+	if stateDirB == stateDirA {
+		if baseB == baseA {
+			stateB = stateA
+		} else {
+			// This should be a rare case, if we hit this we should improve
+			// our normalization function.
+			stateDirB = stateDirB + "_other"
+			stateB, err = client.NewState(stateDirB, baseB)
+			if err != nil {
+				log.Fatalf("ERROR: %s", err)
+			}
+		}
+	} else {
+		stateB, err = client.NewState(stateDirB, baseB)
```

Comment:
@matthewrsj 

This "else" was missing. Probably got wrongly removed when I was shuffling code around to the internal package.

Everything would work fine if both URLs were in the same repo (the thing I was mostly testing lately), hence got unnoticed.